Add config flow and MQTT autodiscover to dsmr_reader integration

Glodenox opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Make the dsmr_reader integration an integration suggestion once a specific MQTT topic is discovered, while also allowing the user to install the dsmr_reader integration via a config flow (there is no real configuration to configure yet, though). This with the purpose of deprecating the configuration.yaml entry.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR is tangentially related to issue #66084 because it is needed if dsmr_reader is to ever have configuration
- This is my very first time coding in Python, please be gentle but also feel free to be strict :)
- This can be triggered by manually publishing one of the MQTT topics to a local broker. The content doesn't really matter.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
-> https://github.com/home-assistant/home-assistant.io/pull/23939

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
-> No new dependencies
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Glodenox committed:
Add config flow and mqtt discovery to dsmr_reader

Glodenox committed:
Include changes created by hassfest script

I thought these would be part of the build or so, but apparently not

Glodenox committed:
Add missing newlines in translation files

homeassistant commented:
Hi @Glodenox, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#71617).

Once you do that we will be able to review and accept this pull request.

Thanks!

Glodenox committed:
Show deprecation message and import config when using configuration.yaml

Glodenox committed:
Update CODEOWNERS file as I apparently never committed that change

Glodenox committed:
Add test for already existing integrations in the user config creation

Glodenox commented:
So it turns out it was actually very easy to show the deprecation message by just adapting the `async_setup_platform` in `sensor.py`. That has been done now, as well as automatically installing a config entry if the integration was put in the configuration.yaml.

Everything went well while testing all the situations one can be in, but there is the one-time situation where the integration is imported from the configuration.yaml and at the same time the discovery via MQTT proposes to add the integration as well. It seems that the MQTT discovery is faster than the import, so I can't immediately think of a good way to prevent that situation. Then again, it won't happen again after a reboot, so it might just be acceptable.

Glodenox commented:
Some more details on why `async_step_import` isn't working: the configuration just adds the platform, but contains no configuration that can be sent to that step. I don't think it causes the issue, but the empty PLATFORM_SCHEMA probably doesn't help either.

Glodenox committed:
Fix isort and flake8 failures

The pre-commit hook failed to trigger in the past due to a pylint bug

Glodenox committed:
Cover missing config flow test coverage section

Glodenox commented:
Well, that last commit should make this pull request completely ready for a review. The workflow would need to be approved again, but locally the test coverage is reported to be 100%.

bdraco commented on homeassistant/components/dsmr_reader/sensor.py:
```diff
@@ -19,7 +24,30 @@ async def async_setup_platform(
+    )
+    current_entries = hass.config_entries.async_entries(domain=DOMAIN)
+    if not current_entries:
+        await hass.config_entries.async_add(
+            ConfigEntry(
+                version=1,
+                domain=DOMAIN,
+                title="DSMR Reader",
+                data={},
+                source=SOURCE_PLATFORM_CONFIG,
+            )
+        )
```
Please implement async_step_import instead

Glodenox replied:
I tried to, but the method didn't trigger at any point. I never really figured out why it didn't work, but I presume it has something to do with there being no configuration other mentioning the dmsr_reader as a sensor. If you're convinced it should be able to work with `async_step_import` as well, I'm willing to give it another shot.

bdraco replied:
You need to start it here. On mobile so I can't put a suggestion but you can 'git grep SOURCE_IMPORT' to find an example

Glodenox replied:
Ah, the fact that I need to trigger it didn't register with me. I've now got a working version locally that does this properly. It also fixed the situation where the integration was imported AND suggested at the same time, so that's great. I'll first check whether it has had any impact on the tests and will then commit.

Glodenox replied:
Just wanted to confirm that the requested changes have been made. I can't remember whether resolving a conversation actually sends out a notification.

Glodenox committed:
Move config import to config_flow.py and trigger it in sensor.py

FransHalsjes commented:
Can you give each sensor a Unique_ID, so sensors can be disabled in HA? I now have lots of "unknown" sensors, all about returns to the grid, which I dont provide. And as they have no unique_id i cannot emove them from my HA dashboard.

Glodenox commented:
> Can you give each sensor a Unique_ID, so sensors can be disabled in HA? I now have lots of "unknown" sensors, all about returns to the grid, which I dont provide. And as they have no unique_id i cannot emove them from my HA dashboard.

I can, and I've even got that code ready locally already 😀 But seeing that pull requests are supposed to be as small as possible and change only what is necessary for the purpose of the request, I'm going to have to await the merging of this one before I can send in a pull request for that.

gjohansson-ST commented on homeassistant/components/dsmr_reader/__init__.py:
```diff
@@ -1 +1,15 @@
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, "sensor")
+    )
+    return True
```
Missing `async_unload_entry`

Glodenox replied:
Ah, that explains why HA was saying it needed to be restarted for the changes to take effect. I've taken the time to also simplify the `async_setup_entry` as it was just doing the work that `async_setup_platforms()` already does for integrations.

gjohansson-ST commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,46 @@
+    ) -> FlowResult:
+        """Handle user step."""
+        if self._async_current_entries():
+            return self.async_show_form(
```
I think it would be better with aborting the flow on this error

Glodenox replied:
That makes more sense, indeed. Adjusted.

bdraco replied:
Is this actually possible since `mqtt` is in `dependencies`?

bdraco replied:
I guess it is if they have no mqtt config and they try to create this manually.

gjohansson-ST commented on homeassistant/components/dsmr_reader/strings.json:
```diff
@@ -0,0 +1,13 @@
+  "config": {
+    "error": {
+      "mqtt_missing": "The MQTT service is required for this integration.",
+      "single_instance_allowed": "[%key:common::config_flow::abort::single_instance_allowed%]"
```
Should be in abort and not in error

Glodenox replied:
Adjusted.

gjohansson-ST reviewed (commented):
Some few comments

Glodenox committed:
Implement async_unload_entry and simplify async_setup_entry

Glodenox committed:
Abort flow instead of showing the flow form when encountering multiple entries

bdraco commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,44 @@
+
+        return self.async_show_form(step_id="user")
+
+    async def async_step_mqtt(self, discovery_info: MqttServiceInfo) -> FlowResult:
+        """Handle a flow initialized by mqtt discovery."""
+        if self._async_in_progress() or self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+        await self.async_set_unique_id(DOMAIN)
+
+        # Offer to register all sensors whenever we encounter at least one dsmr/# topic
+        return await self.async_step_user()
```
If you set `title="DSMR Reader"` in the constructor I think you can drop this since the default likely does what you need https://github.com/home-assistant/core/blob/8af0d91676f118c400037c0354de3e5812cd55d4/homeassistant/helpers/config_entry_flow.py#L112

Glodenox replied:
Inheriting from `DiscoveryFlowHandler` seems to make things easier, indeed. I somewhat fear for the readability as you won't be able to see where the MQTT step gets triggered, so I'll add that information in a comment. I hadn't encountered that class yet.

The only somewhat ugly thing I'm seeing, is that I am then required to provide a discovery function for devices (of which there are none here). I'll probably add a lambda instruction to create a simple Future that always returns false, otherwise I'd need to add a useless function just for that. It would have been nicer if that function was optional.

Glodenox replied:
I've had to add a new private function due to this. Adding it as a lambda or Future didn't seem to be possible. It's a bit of a shame, but it's not the end of the world either. I think I've cleaned up the implementation as much as possible after that change.

Glodenox committed:
Inherit from DiscoveryFlowHandler in the config flow

bdraco commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,47 @@
+    return True
+
+
+@config_entries.HANDLERS.register(DOMAIN)
```
```suggestion
```

Untested, but I'm pretty sure you don't need this

Glodenox replied:
I'm pretty sure I had to add this in the past, as I got an error saying that no handler was configured for the config flow (at https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L747). I've tested it again without that line, and it seems the initialization now properly takes care of adding the handler. It'll be removed in the next commit.

bdraco commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,47 @@
+
+    async def async_step_import(self, _: dict[str, Any] | None) -> FlowResult:
+        """Import from configuration.yaml and create config entry."""
+        # There's no configuration supported for this integration
```
I don't see anything here to guard against creating a new entry each restart

```suggestion
        # There's no configuration supported for this integration
        if self._async_current_entries():
            return self.async_abort(reason="single_instance_allowed")        
```

Glodenox replied:
Good catch! That did indeed seem to happen when I added the integration in the configuration and restarted HA. Somewhat annoyingly, I'm seeing the discovery suggest the integration again while it has already been imported from configuration.yaml (or maybe it's a racing condition?). But it detects the other instance when you try to install it again and doesn't snow up any more afterwards, so I think that's acceptable.

This has now been added.

bdraco commented on tests/components/dsmr_reader/test_config_flow.py:
```diff
@@ -0,0 +1,70 @@
+from tests.common import MockConfigEntry
+
+
+async def test_import_step(hass: HomeAssistant):
+    """Test the import step."""
+    # configure bogus mqtt service to pass validation
+    hass.services.async_register("mqtt", "publish", None)
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_IMPORT},
+    )
+    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+    assert result["title"] == "DSMR Reader"
```
Please add a test to make sure we can only import once

Glodenox replied:
Done. Thanks for catching this!

Glodenox committed:
Single instance check added for config import. Also fixed flake8 and isort issues

bdraco commented:
@depl0y Since you are listed as a codeowner here, is there any chance you can test these changes?

Thanks 👍

Glodenox commented:
I've done some further testing from a user perspective and found some stuff that can be improved, though I haven't seen anything that I'd consider blocking.

**MQTT isn't installed**
- [ ] _configuration.yaml import when MQTT is not installed fails_
    Not a clean error and the integration gets added. Then again, this situation doesn't really make sense and would've given an error in the past as well. I don't think I'll adjust this.
- [ ] _Manual install aborts when MQTT is not installed_
    The error is shown, but it can still be saved. Must be fixed.

**MQTT is installed**
- [x] _configuration.yaml import succeeds_
    Succeeds, but autodetection also happens the first time. It was fixed when I changed to the proper import procedure, but it seems to be back now. Not sure if there's an easy fix, but I haven't found it so far.
- [x] _Autodetection install succeeds_
- [x] _Autodetection doesn't happen more than once_
- [x] _Manual install succeeds_
- [x] _Manual install succeeds even if there are no dsmr topics available (could probably test this)_
       Could be used as a time to inform the user on what to configure. I'll look into this.
- [x] _Manual install can't be done more than once_
- [x] _Autodetection doesn't trigger after install_

Glodenox commented:
> @depl0y Since you are listed as a codeowner here, is there any chance you can test these changes?
> 
> Thanks 👍

Additionally: Wim, I assume you don't mind I added myself to the code owners for this component? I'd like to add some extra configuration and don't mind helping maintain it.

Glodenox committed:
Abort on missing MQTT, improved logging and streamlined tests

bdraco commented:
@Glodenox If the other code owners are not available to test, anyone component to do so who uses this integration is generally fine to do the independent testing. If I use the integration, I can sit in both the review and test seat, but since I don't use this one, generally I like to get another set of 👀 to test before merging.

Glodenox commented:
> FWIW, it is required in this circumstance.
> 
> https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md#rules
Ah, I wasn't aware of that. Makes sense and seems fine to me.

As for the testing: I fully get that. Maybe someone who reported an issue with the integration in the past can do so, but I sort of feel that would be a bit complex to do for someone who isn't used to developing. But then again, this integration could also be tested without having DSMR Reader putting real data on an MQTT server. Only the MQTT server is a necessity for testing. The messages could in theory be added with HA, but that aspect of the integration hasn't changed with this pull request.

balloob commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+        """Import from configuration.yaml and create config entry."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+        if not self.hass.services.has_service(domain="mqtt", service="publish"):
+            _LOGGER.warning("DSMR Reader configuration import failed. MQTT is missing")
+            return self.async_abort(reason="mqtt_missing")
```
MQTT is already a dependency of this integration. The config flow will not start unless MQTT is set up.
```suggestion
```

Glodenox replied:
If I remove this check, I get the error below when someone has the dsmr_reader added as sensor in their configuration.yaml while they don't have MQTT configured. Admittedly, that is a very weird situation to be in, but at the same time it's such an easy thing to catch... I don't mind removing it if other reviewers concur or if I'm overlooking something else.

```
2022-07-02 09:41:41 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up dsmr_reader platform for sensor
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/helpers/entity_platform.py", line 257, in _async_setup_platform
    await asyncio.gather(*pending)
  File "/workspaces/core/homeassistant/helpers/entity_platform.py", line 382, in async_add_entities
    await asyncio.gather(*tasks)
  File "/workspaces/core/homeassistant/helpers/entity_platform.py", line 619, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/workspaces/core/homeassistant/helpers/entity.py", line 809, in add_to_platform_finish
    await self.async_added_to_hass()
  File "/workspaces/core/homeassistant/components/dsmr_reader/sensor.py", line 74, in async_added_to_hass
    await mqtt.async_subscribe(
  File "/workspaces/core/homeassistant/components/mqtt/__init__.py", line 544, in async_subscribe
    async_remove = await hass.data[DATA_MQTT].async_subscribe(
KeyError: 'mqtt'
```

Glodenox replied:
I've removed the mqtt check, as requested. I guess it is unlikely someone wouldn't have mqtt setup when a configuration is being imported anyway.

balloob commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+            return self.async_abort(reason="mqtt_missing")
+        # There's no configuration supported for this integration, so data can be a fixed object
+        return self.async_create_entry(title="DSMR Reader", data={})
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle user step."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+        if not self.hass.services.has_service(domain="mqtt", service="publish"):
+            return self.async_abort(reason="mqtt_missing")
+
+        return await self.async_step_confirm()
```
Can remove because MQTT check not necessary.

balloob replied:
```suggestion
```

Glodenox replied:
When relying on the implementation in DiscoveryFlowHandler without the MQTT check, users can add the component manually while there's no MQTT server installed. I'd prefer to guide users towards configuring the MQTT server first while installing the component, instead of only doing so in documentation. Because almost all logic in this method is already present in the parent class though, I'm thinking of rewriting this method so the MQTT check is the only functional step done within the method (I'd do almost the same thing in the import method):
```
    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Handle user step."""
        if not self.hass.services.has_service(domain="mqtt", service="publish"):
            return self.async_abort(reason="mqtt_missing")

        return await super().async_step_user(user_input)
```

Glodenox replied:
I've removed the mqtt test, as requested. I created issue #75740 to make sure it is documented that an mqtt dependency currently does not trigger an installation when needed.

balloob commented on homeassistant/components/dsmr_reader/strings.json:
```diff
@@ -0,0 +1,13 @@
+  "config": {
+    "abort": {
+      "single_instance_allowed": "[%key:common::config_flow::abort::single_instance_allowed%]",
+      "mqtt_missing": "The MQTT service is required for this integration."
```
```suggestion
```

Glodenox replied:
Removed after getting rid of the two mqtt validations.

balloob commented on homeassistant/components/dsmr_reader/strings.json:
```diff
@@ -0,0 +1,13 @@
+    },
+    "step": {
+      "confirm": {
+        "description": "Make sure to configure the 'split topic' data sources in DSMR Reader, as described [in the documentation](https://www.home-assistant.io/integrations/dsmr_reader#setup)."
```
please insert the link via a `description_placeholder`

Glodenox replied:
Done. I didn't do it at first because then I have to also override the `async_step_confirm` method, but I guess it's not that much of an effort anyway.

Glodenox committed:
Move documentation to placeholder and reuse inherited method logic

Glodenox commented:
I've taken note of https://developers.home-assistant.io/blog/2022/07/08/config_entry_forwards, which will require the replacement of the call to `hass.config_entries.async_setup_platforms`.

Glodenox committed:
Remove MQTT validations again

MQTT not being checked as a dependency is most likely a bug

Glodenox commented:
I've tried to get others to validate the pull request, but it doesn't seem like anyone is up to it. I've already tested it locally multiple times with both a clean environment and a copy of my "production" environment and all works well, so I'd hope that would be good enough.

The MQTT validations I added have been removed again as requested, but I've created issue #75740 to report that an MQTT dependency currently does not trigger the installation/configuration of an MQTT server when needed. I also tried this with existing integrations that have such a dependency and had the same result there.

I would hope this pull request is now ready to be smashed.

@depl0y I'm not sure whether you're busy, but your input would still be welcome 🙂

rouke-broersma commented:
@Glodenox I recently switched from dsmr to dsmr_reader to take advantage of mqtt and I would be willing to test this. Could you give me some direction in how I can test it?

Glodenox commented:
> @Glodenox I recently switched from dsmr to dsmr_reader to take advantage of mqtt and I would be willing to test this. Could you give me some direction in how I can test it?

Oh, that would be nice! Do you already have [a development environment](https://developers.home-assistant.io/docs/development_environment) set up? If not, you can do so following the instructions on that page with the URL of my fork of HA: `https://github.com/Glodenox/core/tree/feature/dsmr-reader-config-flow`. You should then be able to run that version of Home Assistant on your computer, open it at `http://localhost:8123/`, configure your MQTT server, and it should automatically present the option to install the DSMR Reader integration.

rouke-broersma commented:
I'll give it a try right now

jjansen85 commented:
> > @Glodenox I recently switched from dsmr to dsmr_reader to take advantage of mqtt and I would be willing to test this. Could you give me some direction in how I can test it?
> 
> Oh, that would be nice! Do you already have [a development environment](https://developers.home-assistant.io/docs/development_environment) set up? If not, you can do so following the instructions on that page with the URL of my fork of HA: `https://github.com/Glodenox/core/tree/feature/dsmr-reader-config-flow`. You should then be able to run that version of Home Assistant on your computer, open it at `http://localhost:8123/`, configure your MQTT server, and it should automatically present the option to install the DSMR Reader integration.

Tested within dev environment. Behaves like expected, all entities are present.
It shows up automatically (new device detected).

When I start with the old configuration entry, the following message appears:
![Screenshot 2022-08-24 133643](https://user-images.githubusercontent.com/26403565/186409150-45a26b76-5c32-4fef-b3c6-5a451cd5ea4d.png)

DSMR Reader integration is present within my integrations tab.

Glodenox commented:
@rouke-broersma For testing the configuration.yaml migration, you shouldn't have a config entry already configured. Could that explain what you were experiencing? Deleting the integration and restarting HA should be sufficient to make DSMR Reader be listed in the integrations without having configured it via the GUI. At that point, the warning message @jjansen85 mentioned should be shown. Even nicer would be to add a "Repair" action at that point in the new Repairs section. But that would require me to update the branch, which is far from ideal for the reviewing process. I don't think there is anything else that really needs testing here.

@jjansen85 Thanks for testing it out! I didn't expect this pull request to remain open for so long, so I'll try to investigate whether I can just open a pull request for that change without conflicts with this branch.

rouke-broersma commented:
@Glodenox I did 2 tests. One with configuring using configflow and one with migrating from configuration.yaml to configflow. I did not have a configured integration with configflow while testing the configuration.yaml migration. I did not see any migration message but it was late so I will try again later today.

bdraco commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,43 @@
+        if user_input is None:
+            return self.async_show_form(
+                step_id="confirm",
+                description_placeholders={
+                    "documentation_link": "https://www.home-assistant.io/integrations/dsmr_reader#setup"
+                },
```
We can skip this as the UI now has a documentation link built-in
```suggestion
```
<img width="631" alt="Screen_Shot_2022-08-28_at_4_05_55_PM" src="https://user-images.githubusercontent.com/663432/187094515-6df2d3e9-8d00-49d4-9950-cf965367dbc2.png">
n

bdraco commented on homeassistant/components/dsmr_reader/strings.json:
```diff
@@ -0,0 +1,12 @@
+    },
+    "step": {
+      "confirm": {
+        "description": "Make sure to configure the 'split topic' data sources in DSMR Reader, as described [in the documentation]({documentation_link})."
```
```suggestion
        "description": "Make sure to configure the 'split topic' data sources in DSMR Reader."
```

bdraco commented on homeassistant/components/dsmr_reader/translations/en.json:
```diff
@@ -0,0 +1,12 @@
+        },
+        "step": {
+            "confirm": {
+                "description": "Make sure to configure the 'split topic' data sources in DSMR Reader, as described [in the documentation]({documentation_link})."
```
```suggestion
                "description": "Make sure to configure the 'split topic' data sources in DSMR Reader."
```

bdraco commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,43 @@
+
+async def _async_has_devices(_: HomeAssistant) -> bool:
+    """Return true as this integration doesn't support any real devices."""
+    return True
```
This should return False if the data isn't there. I don't have MQTT installed, but manually configuring the user flow would let me set it up.

Glodenox replied:
That would indeed be nicer. The challenge with that is that I'd have to wait for a couple of seconds to await whether any dsmr topics are actually present. I don't think there's a quick and easy way to figure out whether the data is present. Unless I'm overlooking something, of course.

bdraco replied:
`return "mqtt" in hass.config.components` should be a reasonable check here.

Glodenox replied:
Ah, now I see what you mean. I've done some testing and that statement always seems to return true, even in a completely fresh installation. I've gone for `hass.services.has_service(domain="mqtt", service="publish")` instead now, as that can properly distinguish between MQTT being known and being configured.

I've also had to add a string for `config.abort.no_devices_found`. I went for a custom string as "No devices found on the network" doesn't really give a clue on what is going wrong.

bdraco reviewed (dismissed):
Please see comments above.

Thanks 👍

Glodenox committed:
Remove now-unnecessary link to documentation in prompt

The question mark icon in the prompt can now be used to easily access the documentation.

Co-authored-by: J. Nick Koston <nick@koston.org>

Glodenox committed:
Verify MQTT service availability and use FlowResultType

Glodenox commented:
Oh right, that test will fail as the MQTT service won't be marked as available. Should be an easy fix as I have added that in the past. Weird that it didn't show up in my pre-commit hook testing though.

Glodenox committed:
Add test for MQTT presence validation and fix existing test

bdraco commented on tests/components/dsmr_reader/test_config_flow.py:
```diff
@@ -32,6 +53,10 @@ async def test_user_step_with_mqtt(hass: HomeAssistant):
     assert result["step_id"] == "confirm"
     assert result["errors"] is None
 
+    # configure bogus mqtt service to pass flow
+    hass.services.async_register("mqtt", "publish", None)
+    hass.data[DATA_MQTT] = {"async_subscribe": MagicMock()}
```
I think it would be better to check if `mqtt` is in `hass.config.components` since `publish` is an implementation detail of `mqtt` that could get renamed in the future.

```
@core.callback
def async_get_loaded_integrations(hass: core.HomeAssistant) -> set[str]:
    """Return the complete list of loaded integrations."""
    integrations = set()
    for component in hass.config.components:
        if "." not in component:
            integrations.add(component)
            continue
        domain, platform = component.split(".", 1)
        if domain in BASE_PLATFORMS:
            integrations.add(platform)
    return integrations

```

Glodenox replied:
While it seems very unlikely that the publish operation will ever change its name in the context of MQTT, I certainly see your point. The issue is just that `hass.config.components` always seems to contain at least `{'http', 'mqtt'}`, no matter what I do.

It seems like it can be fixed by checking the config entries instead, like with `return "mqtt" in hass.config_entries.async_domains()`. That will tell whether MQTT is configured. I've tested this locally and couldn't find any issues with it.

Glodenox committed:
Verify MQTT with domain instead of service availability

bdraco commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,41 @@
+
+async def _async_has_devices(hass: HomeAssistant) -> bool:
+    """Verify MQTT availability."""
+    return MQTT_DOMAIN in hass.config_entries.async_domains()
```
If MQTT is setup from deprecated platform setup this won't be true.

Glodenox replied:
From reading the mqtt component's code, it would seem like it would create a config entry from the configuration.yaml, which should then get picked up. But I've seen another alternative: checking for the presence of any MQTT data. So I'll replace it with `return DATA_MQTT_CONFIG in hass.data` instead. I'm starting to run out of options here though 😅

bdraco replied:
`return DATA_MQTT_CONFIG in hass.data` is still an implementation detail of `mqtt` that could change.

You've got `mqtt` as a dep here so let's just leave it as `True`

Sorry for the back and forth

Glodenox replied:
No problem! It allowed me to learn some more about the internals of HA. I've reverted the MQTT check.

Glodenox committed:
Revert MQTT configuration check

Glodenox commented:
The changes to the documentation have been prepared, that pull request has been approved.

FutureCow commented:
Would it also be possible to add more than one DSMR-Reader instance to HA?

gjohansson-ST commented on homeassistant/components/dsmr_reader/__init__.py:
```diff
@@ -1 +1,19 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the DSMR Reader integration."""
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```
```suggestion
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```

gjohansson-ST commented on homeassistant/components/dsmr_reader/config_flow.py:
```diff
@@ -0,0 +1,40 @@
+
+
+async def _async_has_devices(_: HomeAssistant) -> bool:
+    """MQTT is set as dependency, so that should be sufficient."""
```
Can we not remove this then?

Glodenox replied:
No. If I remove it, the processing in the parent class fails as it evaluates to false by default, giving the error that no devices were found. I'm open for other solutions to that though, as I don't really like this either.

gjohansson-ST replied:
Ok. See you return the parent class function where it's used.
I guess the better option is too keep it as you have it.

gjohansson-ST commented on homeassistant/components/dsmr_reader/sensor.py:
```diff
@@ -19,7 +23,26 @@ async def async_setup_platform(
 ) -> None:
-    """Set up DSMR Reader sensors."""
+    """Set up DSMR Reader sensors via configuration.yaml and show deprecation warning."""
+    _LOGGER.warning(
```
Change to use repair instead.

Glodenox replied:
It wasn't available when I originally submitted the pull request 😅 But I'll look into implementing it like that, as it would indeed be a lot cleaner. I was planning on doing that in a separate pull request, just to finally get this one merged. But I guess it shouldn't be too much work.

gjohansson-ST replied:
Should be pretty straight forward and there is plenty of examples to just copy.

Glodenox replied:
Done. I just wasn't sure of the value to set in `breaks_in_ha_version`. I assume 2 months (hoping it gets included in the 2022.10 release) would be sufficient for such a minor configuration change.

gjohansson-ST replied:
That's normally how it goes that you set the next version and in case it wouldn't make the cut it's an easy adjustment to the next one. But this looks like good to go now.

gjohansson-ST reviewed (changes requested):
Some last comments

Glodenox commented:
> Would it also be possible to add more than one DSMR-Reader instance to HA?

That would probably only be possible if we support prefixes for the MQTT topics. Those prefixes would then need to be configured in DSMR Reader as well. That seems outside of the scope for this pull request though, and doesn't become impossible after this is merged.

FutureCow commented:
> > Would it also be possible to add more than one DSMR-Reader instance to HA?
> 
> That would probably only be possible if we support prefixes for the MQTT topics. Those prefixes would then need to be configured in DSMR Reader as well. That seems outside of the scope for this pull request though, and doesn't become impossible after this is merged.

Thanks for the reply. I have 2 dsmr-reader instances running. Both report to a different mqtt topic.

Glodenox committed:
Wait for config entry platforms

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

Glodenox committed:
Add a repair issue instead of logging a warning

balloob committed:
hassfest

Glodenox commented:
The PR already exists and was merged half an hour ago: https://github.com/home-assistant/home-assistant.io/pull/23939 I've edited the post to include a link to it.

frenck commented:
LOL 🤦  Handled to many PRs today. Sorry! (and thanks 👍 )
